src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1927,16 +1935,59 @@ private class OffsetAndRule {
             return GetLocalTimeZoneFromWin32Data(timeZoneInformation, dstDisabled);
             
 #else // FEATURE_WIN32_REGISTRY
-            // Without Registry support, just create a dummy TZ for now
+            // Without Registry support, create the TimeZoneInfo from a TZ file
             return GetLocalTimeZoneFromTzFile();
 #endif // FEATURE_WIN32_REGISTRY
         }
 
 
 #if PLATFORM_UNIX
-        static public TimeZoneInfo FindSystemTimeZoneById(string id) {
-            // UNIXTODO
-            throw new NotImplementedException();
+        static public TimeZoneInfo FindSystemTimeZoneById(string id)
+        {
+            if (id == null)
+            {
+                throw new ArgumentNullException("id");
+            }
+            else if (id.Length == 0 || id.Contains("\0"))
+            {
+                throw new TimeZoneNotFoundException(Environment.GetResourceString("TimeZoneNotFound_MissingData", id));
+            }
+
+            // Special case for Utc since there is code that expects only 1 instance of the UTC time zone info to be created.
+            // see CachedData.GetCorrespondingKind()
+            if (string.Equals(id, c_utcId, StringComparison.OrdinalIgnoreCase))
+            {
+                return TimeZoneInfo.Utc;
+            }
+
+            // UNIXTODO: use CachedData
+
+            string timeZoneDirectory = GetTimeZoneDirectory();
+            string timeZoneFilePath = Path.Combine(timeZoneDirectory, id);
+            byte[] rawData;
+            try
+            {
+                rawData = File.ReadAllBytes(timeZoneFilePath);
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                throw new SecurityException(Environment.GetResourceString("Security_CannotReadFileData", id), e);
+            }
+            catch (FileNotFoundException e)
```

Comment:
Might we get a DirectoryNotFoundException in some cases as well?  The catch for IOException will catch DirectoryNotFoundException, but it'll result in the result string InvalidTimeZone_InvalidFileData instead of TimeZoneNotFound_MissingData... I'm assuming the latter would be correct, but I don't actually know.